pinot-broker/src/main/java/org/apache/pinot/broker/routing/v2/RoutingManager.java

```diff
@@ -0,0 +1,481 @@
+    long startTimeMs = System.currentTimeMillis();
+
+    List<ZNRecord> instanceConfigZNRecords =
+        _zkDataAccessor.getChildren(_instanceConfigsPath, null, AccessOption.PERSISTENT);
+    long fetchInstanceConfigsEndTimeMs = System.currentTimeMillis();
+
+    // Calculate new enabled and disabled instances
+    Set<String> enabledInstances = new HashSet<>();
+    List<String> newEnabledInstances = new ArrayList<>();
+    for (ZNRecord instanceConfigZNRecord : instanceConfigZNRecords) {
+      String instance = instanceConfigZNRecord.getId();
+      if (isInstanceEnabled(instanceConfigZNRecord)) {
+        enabledInstances.add(instance);
+
+        // Always refresh the server instance with the latest instance config in case it changes
+        ServerInstance serverInstance = new ServerInstance(new InstanceConfig(instanceConfigZNRecord));
+        if (_enabledServerInstanceMap.put(instance, serverInstance) == null) {
+          newEnabledInstances.add(instance);
+        }
+      }
+    }
+    List<String> newDisabledInstances = new ArrayList<>();
+    for (String instance : _enabledServerInstanceMap.keySet()) {
+      if (!enabledInstances.contains(instance)) {
+        newDisabledInstances.add(instance);
+      }
+    }
+    List<String> changedInstances = new ArrayList<>(newEnabledInstances.size() + newDisabledInstances.size());
+    changedInstances.addAll(newEnabledInstances);
+    changedInstances.addAll(newDisabledInstances);
+    long calculateChangedInstancesEndTimeMs = System.currentTimeMillis();
+
+    // Update routing entry for all tables
+    for (RoutingEntry routingEntry : _routingEntryMap.values()) {
+      try {
+        routingEntry.onInstancesChange(enabledInstances, changedInstances);
+      } catch (Exception e) {
+        LOGGER.error("Caught unexpected exception while updating routing entry on instances change for table: {}",
+            routingEntry.getTableNameWithType(), e);
+      }
+    }
+    long updateRoutingEntriesEndTimeMs = System.currentTimeMillis();
+
+    // Remove new disabled instances from _enabledServerInstanceMap after updating all routing entries to ensure it
+    // always contains the selected instances
+    _enabledServerInstanceMap.keySet().removeAll(newDisabledInstances);
+
+    LOGGER.info(
+        "Processed instance config change in {}ms (fetch {} instance configs: {}ms, calculate changed instances: {}ms, update {} routing entries: {}ms), new enabled instances: {}, new disabled instances: {}",
+        updateRoutingEntriesEndTimeMs - startTimeMs, instanceConfigZNRecords.size(),
+        fetchInstanceConfigsEndTimeMs - startTimeMs, calculateChangedInstancesEndTimeMs - fetchInstanceConfigsEndTimeMs,
+        _routingEntryMap.size(), updateRoutingEntriesEndTimeMs - calculateChangedInstancesEndTimeMs,
+        newEnabledInstances, newDisabledInstances);
+  }
+
+  private static boolean isInstanceEnabled(ZNRecord instanceConfigZNRecord) {
```

Comment:
why is this static?

Reply:
Because it does not rely on any member variable (helper util method)